File: v1/scala/house/src/main/scala/House.scala
class Line(subject: String, verb: String) {
  private lazy val current = s"This is the $subject"
  private lazy val past = s"that $verb the $subject"

  def output(index: Int, count: Int) = {
    val line = if (index == count) current else past
    if (index == 0) line + "." else line
  }
}

object House {
  private val lines = List(
    new Line("house that Jack built", "lay in"),
    new Line("malt", "ate"),
    new Line("rat", "killed"),
    new Line("cat", "worried"),
    new Line("dog", "tossed"),
    new Line("cow with the crumpled horn", "milked"),
    new Line("maiden all forlorn", "kissed"),
    new Line("man all tattered and torn", "married"),
    new Line("priest all shaven and shorn", "woke"),
    new Line("rooster that crowed in the morn", "kept"),
    new Line("farmer sowing his corn", "belonged to"),
    new Line("horse and the hound and the horn", "")
  )

  private def block(index: Int) = (index to (0, -1)).map(i => lines(i).output(i, index)).mkString("\n")

  def rhyme = (0 until lines.length).map(block).mkString("\n\n") + "\n\n"
}

File: prolog/house/house.pl
subject(1, "the house that Jack built").
subject(2, "the malt that lay in").
subject(3, "the rat that ate").
subject(4, "the cat that killed").
subject(5, "the dog that worried").
subject(6, "the cow with the crumpled horn that tossed").
subject(7, "the maiden all forlorn that milked").
subject(8, "the man all tattered and torn that kissed").
subject(9, "the priest all shaven and shorn that married").
subject(10, "the rooster that crowed in the morn that woke").
subject(11, "the farmer sowing his corn that kept").
subject(12, "the horse and the hound and the horn that belonged to").

combine(Part, "", Part) :- !.
combine(Part, Acc, Result) :- format(string(Result), "~s ~s", [Part, Acc]).

verse(N, Lyric) :-
    numlist(1, N, Ns),
    maplist(subject, Ns, Parts),
    foldl(combine, Parts, "", Subjects),
    format(string(Lyric), "This is ~s.", [Subjects]).

recite(Verse, Lyrics) :- recite(Verse, Verse, Lyrics).
recite(From, To, Lyrics) :-
    numlist(From, To, Ns),
    maplist(verse, Ns, Lyrics).
